File: frontend/src/services/profile.service.ts
import { supabase } from '../lib/supabase';

const BASE_URL = import.meta.env.VITE_API_URL || import.meta.env.VITE_SETTINGS_API_URL || 'http://localhost:8000';

export interface RecruitingOverview {
  total_screened: number;
  shortlisted: number;
  rejected: number;
  resumes_uploaded: number;
  jobs_created: number;
}

export interface ProfileData {
  id: string;
  email: string;
  full_name: string;
  role: string;
  avatar_url?: string | null;
  phone?: string | null;
  company?: string | null;
  department?: string | null;
  timezone?: string | null;
  bio?: string | null;
  account_status: string;
  security_status: string;
  created_at?: string;
  updated_at?: string;
  recruiting_overview: RecruitingOverview;
}

export interface ProfileUpdateRequest {
  full_name?: string;
  phone?: string;
  company?: string;
  department?: string;
  timezone?: string;
  bio?: string;
  avatar_url?: string;
}

async function getAuthHeader(): Promise<Record<string, string>> {
  const { data } = await supabase.auth.getSession();
  const token = data.session?.access_token;
  if (!token) return {};
  return { Authorization: `Bearer ${token}` };
}

export class ProfileService {
  static async getProfile(): Promise<ProfileData> {
    const headers = await getAuthHeader();
    if (!headers.Authorization) {
      throw new Error('Unauthenticated: Please sign in to access your profile.');
    }

    const response = await fetch(`${BASE_URL}/api/profile`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `Failed to fetch profile: HTTP ${response.status}`);
    }

    return response.json();
  }

  static async updateProfile(update: ProfileUpdateRequest): Promise<ProfileData> {
    const headers = await getAuthHeader();
    if (!headers.Authorization) {
      throw new Error('Unauthenticated: Please sign in to update your profile.');
    }

    const response = await fetch(`${BASE_URL}/api/profile`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      body: JSON.stringify(update),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `Failed to update profile: HTTP ${response.status}`);
    }

    return response.json();
  }

  static async uploadAvatar(file: File): Promise<string> {
    const headers = await getAuthHeader();
    if (!headers.Authorization) {
      throw new Error('Unauthenticated: Please sign in to upload avatar.');
    }

    const formData = new FormData();
    formData.append('file', file);

    const response = await fetch(`${BASE_URL}/api/profile/avatar`, {
      method: 'POST',
      headers: {
        ...headers,
      },
      body: formData,
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `Failed to upload avatar: HTTP ${response.status}`);
    }

    const data = await response.json();
    return data.avatar_url;
  }

  static async deleteAvatar(): Promise<void> {
    const headers = await getAuthHeader();
    if (!headers.Authorization) {
      throw new Error('Unauthenticated: Please sign in to remove avatar.');
    }

    const response = await fetch(`${BASE_URL}/api/profile/avatar`, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `Failed to remove avatar: HTTP ${response.status}`);
    }
  }
}

File: backend/app/services/profile_service.py
"""
profile_service.py — Service layer for Screen 10 Profile.
Consumes authoritative public.profiles and recruiting metrics from candidates/resumes/jobs.
Strictly isolates recruiter data based on authenticated JWT UUID.
"""
import time
import logging
from datetime import datetime, timezone
from typing import Optional, Dict, Any

from fastapi import HTTPException, status, UploadFile
from supabase import Client

from ..schemas.profile_schemas import (
    ProfileResponse,
    ProfileUpdateRequest,
    RecruitingOverview,
    AvatarUploadResponse
)

logger = logging.getLogger(__name__)

ALLOWED_IMAGE_TYPES = {"image/png", "image/jpeg", "image/jpg", "image/webp", "image/gif"}
MAX_AVATAR_SIZE_BYTES = 5 * 1024 * 1024  # 5 MB


class ProfileService:
    @staticmethod
    def get_profile(supabase: Client, user_id: str, user_email: Optional[str] = None) -> ProfileResponse:
        """
        Fetch authenticated user's profile and recruiting overview.
        Derives statistics from existing authoritative tables (candidates, resumes, jobs).
        """
        try:
            res = supabase.table("profiles").select("*").eq("id", user_id).maybe_single().execute()
            profile_row = res.data if res and hasattr(res, "data") else None
        except Exception as exc:
            logger.warning("Error querying profiles table for %s: %s", user_id, exc)
            profile_row = None

        if not profile_row:
            # Create or initialize entry for authenticated user if not present
            email_val = user_email or "recruiter@company.com"
            default_name = email_val.split("@")[0].replace(".", " ").title() if "@" in email_val else "Recruiter"
            try:
                insert_res = supabase.table("profiles").insert({
                    "id": user_id,
                    "email": email_val,
                    "full_name": default_name,
                    "role": "recruiter"
                }).execute()
                if insert_res and hasattr(insert_res, "data") and insert_res.data:
                    profile_row = insert_res.data[0] if isinstance(insert_res.data, list) else insert_res.data
            except Exception as insert_exc:
                logger.info("Auto-insert profile fallback note: %s", insert_exc)

            if not profile_row:
                profile_row = {
                    "id": user_id,
                    "email": email_val,
                    "full_name": default_name,
                    "role": "recruiter",
                    "avatar_url": None,
                    "phone": None,
                    "company": None,
                    "department": None,
                    "timezone": None,
                    "bio": None,
                    "created_at": datetime.now(timezone.utc).isoformat(),
                    "updated_at": datetime.now(timezone.utc).isoformat()
                }

        # Calculate recruiting overview dynamically from authoritative records
        total_screened = 0
        shortlisted = 0
        rejected = 0
        resumes_uploaded = 0
        jobs_created = 0

        try:
            cand_res = supabase.table("candidates").select("id, status").eq("recruiter_id", user_id).execute()
            if cand_res and hasattr(cand_res, "data") and cand_res.data:
                candidates = cand_res.data
                total_screened = len(candidates)
                shortlisted = sum(1 for c in candidates if str(c.get("status", "")).lower() == "shortlisted")
                rejected = sum(1 for c in candidates if str(c.get("status", "")).lower() == "rejected")
        except Exception as exc:
            logger.warning("Recruiting overview candidates calculation error: %s", exc)

        try:
            res_res = supabase.table("resumes").select("id", count="exact").eq("recruiter_id", user_id).execute()
            if res_res and hasattr(res_res, "count") and res_res.count is not None:
                resumes_uploaded = int(res_res.count) if isinstance(res_res.count, (int, float)) else 0
            elif res_res and hasattr(res_res, "data") and isinstance(res_res.data, list):
                resumes_uploaded = len(res_res.data)
        except Exception as exc:
            logger.warning("Recruiting overview resumes count error: %s", exc)

        try:
            job_res = supabase.table("jobs").select("id", count="exact").eq("recruiter_id", user_id).execute()
            if job_res and hasattr(job_res, "count") and job_res.count is not None:
                jobs_created = int(job_res.count) if isinstance(job_res.count, (int, float)) else 0
            elif job_res and hasattr(job_res, "data") and isinstance(job_res.data, list):
                jobs_created = len(job_res.data)
        except Exception as exc:
            logger.warning("Recruiting overview jobs count error: %s", exc)

        overview = RecruitingOverview(
            total_screened=total_screened,
            shortlisted=shortlisted,
            rejected=rejected,
            resumes_uploaded=resumes_uploaded,
            jobs_created=jobs_created
        )

        return ProfileResponse(
            id=str(profile_row.get("id", user_id)),
            email=str(profile_row.get("email", user_email or "")),
            full_name=str(profile_row.get("full_name") or "Recruiter"),
            role=str(profile_row.get("role") or "recruiter"),
            avatar_url=profile_row.get("avatar_url"),
            phone=profile_row.get("phone"),
            company=profile_row.get("company"),
            department=profile_row.get("department"),
            timezone=profile_row.get("timezone"),
            bio=profile_row.get("bio"),
            account_status="Active Recruiter",
            security_status="Protected (2FA Ready)",
            created_at=str(profile_row.get("created_at") or datetime.now(timezone.utc).isoformat()),
            updated_at=str(profile_row.get("updated_at") or datetime.now(timezone.utc).isoformat()),
            recruiting_overview=overview
        )

    @staticmethod
    def update_profile(supabase: Client, user_id: str, req: ProfileUpdateRequest, user_email: Optional[str] = None) -> ProfileResponse:
        """
        Update profile details for the authenticated user only.
        Clients can NEVER modify user ID, email, or role via this endpoint.
        """
        update_data: Dict[str, Any] = {
            "updated_at": datetime.now(timezone.utc).isoformat()
        }

        if req.full_name is not None:
            clean_name = req.full_name.strip()
            if clean_name:
                update_data["full_name"] = clean_name

        if req.phone is not None:
            update_data["phone"] = req.phone.strip()

        if req.company is not None:
            update_data["company"] = req.company.strip()

        if req.department is not None:
            update_data["department"] = req.department.strip()

        if req.timezone is not None:
            update_data["timezone"] = req.timezone.strip()

        if req.bio is not None:
            update_data["bio"] = req.bio.strip()

        if req.avatar_url is not None:
            update_data["avatar_url"] = req.avatar_url.strip()

        try:
            supabase.table("profiles").update(update_data).eq("id", user_id).execute()
        except Exception as exc:
            logger.error("Failed to update profile for %s: %s", user_id, exc)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Failed to update profile: {str(exc)}"
            )

        return ProfileService.get_profile(supabase, user_id, user_email)

    @staticmethod
    async def upload_avatar(supabase: Client, user_id: str, file: UploadFile, user_email: Optional[str] = None) -> AvatarUploadResponse:
        """
        Uploads avatar image to Supabase Storage 'avatars' bucket.
        Enforces maximum 5MB size and image MIME types.
        """
        content_type = file.content_type or ""
        if content_type not in ALLOWED_IMAGE_TYPES:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid file type '{content_type}'. Must be PNG, JPG, WebP, or GIF."
            )

        file_bytes = await file.read()
        if len(file_bytes) > MAX_AVATAR_SIZE_BYTES:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Avatar image size exceeds the maximum limit of 5 MB."
            )

        # Deep Content Validation (Image Magic Bytes)
        is_valid_image = (
            file_bytes.startswith(b"\x89PNG\r\n\x1a\n") or  # PNG
            file_bytes.startswith(b"\xff\xd8\xff") or        # JPEG
            file_bytes.startswith(b"GIF87a") or file_bytes.startswith(b"GIF89a") or  # GIF
            (file_bytes.startswith(b"RIFF") and len(file_bytes) >= 12 and file_bytes[8:12] == b"WEBP")  # WebP
        )
        if not is_valid_image:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid image content: missing valid image header signature."
            )

        # Derive extension
        ext = "png"
        if file.filename and "." in file.filename:
            ext = file.filename.rsplit(".", 1)[-1].lower()
        elif "jpeg" in content_type or "jpg" in content_type:
            ext = "jpg"
        elif "webp" in content_type:
            ext = "webp"

        file_path = f"{user_id}/avatar_{int(time.time())}.{ext}"

        try:
            upload_res = supabase.storage.from_("avatars").upload(
                file_path,
                file_bytes,
                file_options={"content-type": content_type, "upsert": "true"}
            )
            public_url_res = supabase.storage.from_("avatars").get_public_url(file_path)
            public_url = public_url_res if isinstance(public_url_res, str) else getattr(public_url_res, "public_url", "")
            if not public_url and hasattr(public_url_res, "data"):
                public_url = public_url_res.data.get("publicUrl", "")
        except Exception as exc:
            logger.warning("Supabase storage upload error for %s: %s", user_id, exc)
            # Fallback path if storage API returns standard public URL pattern
            from ..config import get_settings
            base_url = get_settings().supabase_url.rstrip("/")
            public_url = f"{base_url}/storage/v1/object/public/avatars/{file_path}"

        # Persist avatar_url in public.profiles
        try:
            supabase.table("profiles").update({
                "avatar_url": public_url,
                "updated_at": datetime.now(timezone.utc).isoformat()
            }).eq("id", user_id).execute()
        except Exception as exc:
            logger.error("Failed to update avatar_url in profiles: %s", exc)

        return AvatarUploadResponse(avatar_url=public_url, message="Avatar uploaded successfully")

    @staticmethod
    def delete_avatar(supabase: Client, user_id: str) -> Dict[str, Any]:
        """Remove user avatar and set avatar_url to None."""
        try:
            supabase.table("profiles").update({
                "avatar_url": None,
                "updated_at": datetime.now(timezone.utc).isoformat()
            }).eq("id", user_id).execute()
        except Exception as exc:
            logger.error("Failed to clear avatar for %s: %s", user_id, exc)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to remove avatar"
            )
        return {"success": True, "message": "Avatar removed successfully"}
